### ISSUE TITLE
Mounted volumes on Docker containers should be accessible from the Host VM

### ISSUE THREAD

**Issue by xcoulon** [projectatomic/adb-atomic-developer-bundle#118]
Mounted volumes on Docker containers should be accessible from the Host VM, so a user can deploy his application from his IDE on the Host VM into the mounted volume (ie, a specific directory in the host VM) and the application server in the Docker container will be able to pick the changes automatically.

**Comment by navidshaikh** [comment]
in the scope of vagrant-adbinfo plugin

**Comment by bexelbie** [comment]
@praveenkumar @navidshaikh and @bexelbie are to explore SSHFS for this for all platforms

**Comment by LalatenduMohanty** [comment]
@navidshaikh @praveenkumar @bexelbie I came across https://github.com/winnfsd/vagrant-winnfsd . Need to evaluate if it is useful for us

**Comment by maxandersen** [comment]
One important aspect of mounting is that its best if you can use the same paths as the underlying system.

i.e. "/Users/max/myapp/data" and not "/netfs/myapp/data" 

Dockermachine solves that by mounting users home dir in the same path as what the host will be.

On windows this will be tricky due to Drive part of paths and the use of \ vs /, but should at least have the path name be consistent imo.

**Comment by tuxdna** [comment]
In current state of things, by default a `/vagrant`  ( using `rsync ./  user@HOST:/vagrant` ) folder is created . However this is a one time sync ( it is not really a filesystem mount ).

@xcoulon I want to confirm what you mean by the mounts, which enable IDE  --> Docker deployment.

```
HOST Machine:

/var/path_on_host/file.txt

VM:

/var/path_on_vm/file.txt

Docker Container:

/var/path_on_container/file.txt

```

So if we change `file.txt` in Host VM, it should get reflected in Docker Container.

It seems like we can do it two ways:
1.   `HOST Machine -> VM -> Docker Container` 
2.   `VM -> Docker Container` +  `VM -> HOST Machine`

Option 2 doesn't really make sense because, I would think that all the developer's source code originates from `HOST Machine`. So if we mount a folder from VM into HOST Machine, the code would persist inside in the VM, and not on the HOST Machine ( only visible, while the mount exists ).

@dustymabe , @maxandersen : Suggestions?

**Comment by raghavendra-talur** [comment]
Guys, 
Looking at the requirements it looks like rsync-auto would be enough for you.  Did you try that?

**Comment by bexelbie** [comment]
@raghavendra-talur auto-rsync would copy an entire tree, say /Users on a Mac, every time anything was updated.  A mount smells cleaner to me.

**Comment by raghavendra-talur** [comment]
@bexelbie makes sense.

I was thinking it was just the project folder and anyhow rsync is intelligent enough to copy only new data and not update the mtime for old files.
But you are right, it there is lot of data it would make sense to have a mount and not a copy.

**Comment by concaf** [comment]
Two way sync from host to guest:

Solution 1 - SSHFS (vagrant-sshfs plugin)

Pros:
- One solution for all platforms

Cons:
- No longer maintained, so it has to be maintained by us since it might break with future Vagrant versions
- Performance issues (high latency, reliable over heavy load?) since it's working over SSH
- SSH server has to be packaged with the Windows installer

Solution 2 - [NFS](https://www.vagrantup.com/docs/synced-folders/nfs.html) (Linux + Mac) and [SMB](https://www.vagrantup.com/docs/synced-folders/smb.html) (Windows)

Pros:
- Maintained by Vagrant, so easier to maintain. Also, this will have more testing and user feedback backing it than our customized vagrant-sshfs plugin will have.
- Works out of the box, just one `vagrant up`
- No tinkering required on the Windows host (but NFS server required on Linux and Mac hosts)

Cons:
- The Vagrantfile gets a bit typical to understand, but can we also make this solution into a plugin (don't know how that works)?
- Multiple solutions for one problem, but both of these are well built in and their implementation (mounts) is automated and hidden away from the end user, so no change in the user experience, rather it is more seamless
- Vagrant is not 'supposed' to maintain these in the future, but if we have to maintain the solution, then why not this than vagrant-sshfs?

I have tried out the second solution ([Vagrantfile](https://raw.githubusercontent.com/containscafeine/dots/master/Vagrantfile)) (please change the mount points according to your environment)

This Vagrantfile will not work out of the box since Vagrant does not provision before mounting the synced folders, but it should not matter if cifs-utils is packaged inside the adb vagrant box, so if someone wants to try it:

Steps for Windows:
- `vagrant up`
- `vagrant provision`
- `vagrant reload`

Steps for Linux:
- Allow incoming connections to the NFS server on the host
- `vagrant up`

Please correct me if I did not understand the issue correctly or stated something off track.

**Comment by tuxdna** [comment]
SSHFS is working on Linux and Window as of this commit - https://github.com/tuxdna/vagrant-sshfs/commit/9f0d2c9c2dff3f65c4dd3f6f6c10891bee8fecaf

However there are some steps involved like updating /etc/fuse.conf and setting up Windows SSH Server.

**Comment by navidshaikh** [comment]
> SSHFS is working on Linux and Window as of this commit - tuxdna/vagrant-sshfs@9f0d2c9

Nice!

> However there are some steps involved like updating /etc/fuse.conf and setting up Windows SSH Server.

can we document this ?

**Comment by concaf** [comment]
I am running some file system benchmarks (bonnie++) to test if performance is an issue.
Environment: projectatomic/adb, Memory: 1839, no CPU specified in Vagrantfile

[Results for SSHFS](https://gist.githubusercontent.com/containscafeine/e19a7eb838539ac45164/raw/aff05ea65c3209b1819b8963208b5a736ce9540a/SSHFS%2520bonnie%2520-%2520Linux)
[Results for NFS](https://gist.githubusercontent.com/containscafeine/b2ac3c1bc6093ddc1884/raw/9621a92235696f6e6cde8935f4092164dce7f4c7/NFS%2520bonnie%2520-%2520Linux)

To get a better understanding of the output, read [here](http://www.coker.com.au/bonnie++/readme.html)

I will try and format the results a bit,

Dataset used:
- NFS: 4GB
- SSHFS: 4GB

Sequential Output (K/sec):
- NFS: Per-Character 1747, Block 211632, Rewrite 100980 
- SSHFS: Per-Character 39, Block 13465, Rewrite 7809

Sequential Input (K/sec):
- NFS: Per-Character 3869, Block 307696
- SSHFS: Per-Character 3472, Block 14925

Random Seeks: (/sec)
- NFS: 3246
- SSHFS: 1283

The same behavior continues for other checks, testing in Windows is in progress.

The latency for SSHFS is about 50-100 times more than latency in NFS, please refer to the files above.

P.S. For an overall picture, this is how `dd` works on both.

NFS:
`[vagrant@centos7-adb ~]$ dd if=/dev/urandom of=/home/vagrant/adb_mount/dd.out bs=4M count=100
100+0 records in
100+0 records out
419430400 bytes (419 MB) copied, 34.9642 s, 12.0 MB/s`

SSHFS:
`[vagrant@centos7-adb ~]$ dd if=/dev/urandom of=/home/vagrant/adb_mount/dd.out bs=4M count=100
100+0 records in
100+0 records out
419430400 bytes (419 MB) copied, 60.9014 s, 6.9 MB/s`

So SSHFS takes twice the time as NFS.

@navidshaikh @dustymabe @tuxdna @xcoulon @maxandersen @LalatenduMohanty 
Will this not be an issue for initial setup or huge codebase or while working with media files or databases (depends on the architecture of the application)?
Keeping the ideal scenario aside, we don't know how the end user will be using the box!

Please correct me I'm speaking gibberish somewhere.

**Comment by tuxdna** [comment]
@containscafeine thats a super impressive result. SSHFS is surely a bad option when performance is a concern ( big files / huge number of files / high IO ).

**Comment by concaf** [comment]
Windows:

tl;dr - I cannot make out much from the following results, but my best 'guess' is that SSHFS and SMB are actually comparable on Windows, but SMB is always a tad bit faster.

Note: Linux benchmarks were done on an SSD drive, but my Windows machine does not have it. However, it should not matter when comparing SSHFS and SMB since the environment is the same.

Environment: projectatomic/adb, Memory: 1839, CPUs: 2

`dd` output:

SMB:
`[vagrant@centos7-adb ~]$ dd if=/dev/urandom of=/vagrant/dd.out bs=4M count=100
100+0 records in
100+0 records out
419430400 bytes (419 MB) copied, 39.9931 s, 10.5 MB/s`

SSHFS:
`[vagrant@centos7-adb ~]$ dd if=/dev/urandom of=/home/vagrant/adb_mount/dd.out bs=4M count=100
100+0 records in
100+0 records out
419430400 bytes (419 MB) copied, 44.3119 s, 9.5 MB/s`

Detailed results:
- [SSHFS (bonnie)](https://gist.githubusercontent.com/containscafeine/d659f4e32d757472299d/raw/e60385d3d1df9975adfb8cbd7032f2a2145ff61f/SSHFS%2520bonnie%2520-%2520Windows)
- [SMB (bonnie)](https://gist.githubusercontent.com/containscafeine/7fd5b09310bbfb2f06e5/raw/515350124d98c135585e6879d1bce01a5181017c/SMB%2520bonnie%2520-%2520Windows)

**Comment by concaf** [comment]
@tuxdna, yes it's true but as we discussed, we need way more tests and in a much more automated and sane environments. I hope these results do not convey any wrong results, because the sample size was just 1, we need more of these to arrive at a conclusion. But this should do for a rough idea, right?

**Comment by bexelbie** [comment]
@containscafeine can you update this with a link to the mail to container-tools that contains the results of our conversation?

**Comment by concaf** [comment]
@bexelbie [here](https://www.redhat.com/archives/container-tools/2016-February/msg00108.html) is the link if someone wants to follow the discussion on the mailing list.
There were no results to the thread, but it was agreed upon over other channels that SSHFS is the way to go since performance does not seem to be the primary concern right now, and except for that SSHFS fits the bill for all platforms.

**Comment by concaf** [comment]
@maxandersen @tuxdna @LalatenduMohanty
I have tried [this](https://www.redhat.com/archives/container-tools/2016-February/msg00172.html) solution by @dustymabe on projectatomic/adb vagrant box, on Windows 10 and Fedora 23, and it works fine for me.
Have you guys tried it on OS X/Windows/Linux yet?

**Comment by dustymabe** [comment]
New code is in http://github.com/dustymabe/vagrant-sshfs. Instructions are there. Please try it out and give feedback.

**Comment by dustymabe** [comment]
New release of vagrant-sshfs is out. Looks like we are now seriously evaluating this as an option.

**Comment by LalatenduMohanty** [comment]
@maxandersen @hferentschik Closing this issue as vagrant-sshfs fixes the overall issue. For vagrant-sshfs specific issues we should file new issues.

**Comment by hferentschik** [comment]
I don't think this is really resolved yet. Also we don't have anything yet in the _Vagrantfile_ which actually does volume mounts. But I guess we can open a new issue.

**Comment by hferentschik** [comment]
Here we go - https://github.com/projectatomic/adb-atomic-developer-bundle/issues/339

**Comment by hferentschik** [comment]
Actually, there was already an issue - https://github.com/projectatomic/adb-atomic-developer-bundle/issues/336
